Contrib/FreeWilson/freewilson.py

```diff
@@ -0,0 +1,486 @@
+INFO:root:	Core	1
+INFO:root:	R1	73
+INFO:root:	R10	2
+INFO:root:	R3	445
+100%|███████████████████████████████████████████████████████| 64970/64970 [00:05<00:00, 11247.38it/s]
+INFO:root:Wrote 3 results out of 64970
+	In Training set: 628
+	Bad MW: 0
+	Bad Pred: 64339
+	Bad Filters: 0
+	Bad smi: 0
+	min mw: 380.429
+	max mw: 772.4030000000001
+	
+	min pred: 2.8927324757327666
+	max pred: 8.148182660251193
+```
+
+Using FMCS to find a scaffold
+-----------------------------
+
+If you don't have a good feel for the dataset, you can generate the scaffold by using
+the rdFMCS package.  The following tries to find a decent MCS that covers 80% of the
+input structures
+
+
+```
+>>> from rdkit.Chem import rdFMCS
+>>> mcs = rdFMCS.FindMCS(mols, threshold=0.8, atomCompare=rdFMCS.AtomCompare.CompareAny,
+...                      completeRingsOnly=True)
+>>> decomp = FWDecompose(mcs.queryMol, mols, scores)
+```
+
+Using Molecule filters
+----------------------
+Finally, the molecule filter can be used to prune greasy or otherrwise undesirable
+molecules:
+
+```
+from rdkit.Chem import Descriptors
+>>> for pred in fw.FWBuild(free, mol_filter=lambda mol: -3 < Descriptors.MolLogP(mol) < 3):
+...   print(pred.smiles, pred.prediction)
+"""
+
+from rdkit.Chem import rdRGroupDecomposition as rgd
+from rdkit.Chem import molzip, Descriptors
+from rdkit import rdBase
+from rdkit import Chem
+from sklearn.linear_model import Ridge
+from sklearn.metrics import r2_score
+from collections import defaultdict, namedtuple
+from tqdm import tqdm
+import itertools
+import math
+import sys
+from typing import List
+import logging
+import csv
+import re
+from typing import Generator
+logger = logging.getLogger("rdkit.Chem.FreeWilson")
+
+FreeWilsonPrediction = namedtuple("FreeWilsonPrediction", ['prediction', 'smiles', 'rgroups'])
+
+# hydrogens with dummy atoms
+hspat = re.compile(r"\[[H0-9]+\]\[\*:([0-9]+)")
+# all dummy atoms
+dummypat = re.compile(r"\*:([0-9]+)")
+
+
+class RGroup:
+    """FreeWilson RGroup
+        smiles - isomeric smiles of the rgroup
+        rgroup - Rgroup name ['Core', 'R1', 'R2'...]
+        count - number of molecules with the rgroup
+        coefficient - ridge regression coefficient
+        idx - one-hot encoding for the rgroup
+    """
+    def __init__(self, smiles, rgroup, count, coefficient, idx=None):
+        self.smiles = smiles           # smiles for the sidechain (n.b. can be a core as well)
+        self.rgroup  = rgroup          # rgroup Core, R1, R2,...
+        self.count = count             # num molecules with this rgruop
+        self.coefficient = coefficient # ridge coefficient
+        self.idx = None                # descriptor index
+        hs = [int(x) for x in hspat.findall(smiles)]
+        assert len(hs) <= 1
+        self.hs = hs                   # hydrogen with dummy atoms    
+        self.dummies = tuple([int(x) for x in sorted(dummypat.findall(smiles))])
+
+        # Assemble some additive properties 
```

Comment:
Since this is only being used as a filter in the enumeration phase, how about just using the number of heavy atoms? That's easier to calculate (no worries about sanitization) and not subject to "anti-bromine bias" ;-)

Reply:
That certainly could be added, but I think the concept of filter by molecular weight its pretty ingrained.

Reply:
Sure. But that doesn't mean that it's the best thing to use. :-)

Reply:
I'll go ahead and add the num heavy atoms so we can slowly start changing the world...